File: P56 Merge Intervals/Solution.java
class Solution {
    // Time: o(nlogn)
    // Space: O(n)
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) {
            return intervals;
        }
        Arrays.sort(intervals, (i1, i2) -> Integer.compare(i1[0], i2[0]));
        List<int[]> result = new ArrayList<>();
		int[] newInterval = intervals[0];
		result.add(newInterval);
        for (int[] interval: intervals) {
            if (newInterval[1] >= interval[0]) {
				newInterval[1] = Math.max(newInterval[1], interval[1]);
		    } else {                           
				newInterval = interval;
                result.add(newInterval);
			} 
        }
        return result.toArray(new int[result.size()][]);
    }
}

File: P56 Merge Intervals/solution.py
class Solution:
    # Time: o(nlogn)
    # Space: O(n)
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 1:
            return intervals
        intervals = sorted(intervals,key=lambda l:l[0])
        newInterval = intervals[0]
        result = [newInterval]
        for interval in intervals:
            if newInterval[1] >= interval[0]:
                newInterval[1] = max(newInterval[1], interval[1])
            else:
                newInterval = interval
                result.append(newInterval)
        return result